### PR TITLE
Pykd target

### PR DESCRIPTION
**Pull request by dark-lbp**
Add pykd target support

### CODE REVIEW

File: katnip/targets/pykd_dbg.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+        pykd.deinitialize()
+        super(PykdTarget, self).post_test(test_num)
+
+    def _send_to_target(self, data):
+        '''
+
+        :param data: mutational_data (might by file path)
+        '''
+        self.logger.info('send called')
+        '''
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: This should be a comment, or removed, not a string

File: katnip/targets/pykd_dbg.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    def post_test(self, test_num):
+        if self.is_crash.wait(self._timeout):
+            self.crash_dump_finished.wait(self._timeout)
+        self._stop_process()
+        # for each test case we need deinitialize the pykd, other wise event handle will not work.
+        pykd.deinitialize()
+        super(PykdTarget, self).post_test(test_num)
+
+    def _send_to_target(self, data):
+        '''
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: mutational_data is not a got description, why not use the original description?
typo - might by --> might be

File: katnip/targets/pykd_dbg.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+        :param data: mutational_data (might by file path)
+        '''
+        self.logger.info('send called')
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: better to use self.logger.debug() in most cases.

File: katnip/targets/pykd_dbg.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    '''
+    This class is used to Handler Event from PYKD, you can overwrite any methods which listed in
+    this link(https://pykd.codeplex.com/wikipage?title=PYKD%200.3.%20API%20Reference#eventHandler)
+    '''
+    def __init__(self, target):
+        '''
+        :param target: Our PykdTarget
+        '''
+        super(EventHandler, self).__init__()
+        self._Target = target
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: _Target -> _target

File: katnip/targets/pykd_dbg.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+            Will run ``/tmp/myApp -a -c mutational_data`` using pykd for evey mutation with timeout of 2 seconds
+
+        '''
+        super(PykdTarget, self).__init__(name, logger)
+        assert(process_path)
+        assert(os.path.exists(process_path))
+        self._process_path = process_path
+        self._process_name = os.path.basename(process_path)
+        self._process_args = process_args
+        self.process_data = None
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py:122]: process_data -> _process_data

File: katnip/targets/pykd_dbg.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        self._break_points = break_points
+        self._handler = handler
+        self.handler = None
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: handler -> _handler

File: katnip/targets/pykd_dbg.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        self._timeout = timeout
+        self._process = None
+        self._pid = None
+        self._system_pid = None
+        self.is_crash = threading.Event()
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py:129]: is_crash -> _is_crash

File: katnip/targets/pykd_dbg.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        self.crash_dump_finished = threading.Event()
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py:130]: crash_dump_finished -> _crash_dump_finished

File: katnip/targets/pykd_dbg.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+                pykd.go()
+        except:
+            self.logger.error('Got an exception in _debug_server')
+            self.logger.error(traceback.format_exc())
+
+    def _start_server_thread(self):
+        '''
+        start the server thread
+        '''
+        self.server_thread = FuncThread(self._debug_server)
```
Comment (BinyaminSharet) [inline katnip/targets/pykd_dbg.py]: server_thread -> _server_thread

**Review by BinyaminSharet — CHANGES_REQUESTED** [review]
I can't really comment on most things as I'm not familiar with pykd, but there are few minor changes -
1. member names should be lower case and if they are private start with _.
2. Use self.logger.debug instead of self.logger.info whenever possible.

Cheers :)

File: katnip/targets/pykd_dbg.py (hunk 5 above)
Comment (dark-lbp) [reply katnip/targets/pykd_dbg.py:122]: hmm, it's used in my custom EventHandler, since sometime we need to copy the file or data to some other folder.

Comment (BinyaminSharet) [reply katnip/targets/pykd_dbg.py:122]: if it is used then it's fine :smile:

File: katnip/targets/pykd_dbg.py (hunk 7 above)
Comment (dark-lbp) [reply katnip/targets/pykd_dbg.py:129]: This is also used in EventHandler.

Comment (BinyaminSharet) [reply katnip/targets/pykd_dbg.py:129]: all right, I tried to comment each member so you won't forget any, but just fix the relevant ones :smile:

File: katnip/targets/pykd_dbg.py (hunk 2 above)
Comment (dark-lbp) [reply katnip/targets/pykd_dbg.py]: Like this one?
```python
    def _send_to_target(self, payload):
        '''

        :param payload: payload to send (might be file path)
        '''
        self.logger.info('send called')
        '''
        starting debug server
        '''
        self.process_data = [payload]
        self._start_server_thread()
```

**Review by BinyaminSharet — APPROVED** [review]

**Pull request merged**